### PR TITLE
WIP: Add synchronization in lite verify

### PR DESCRIPTION
**Pull request by HaoyangLiu**
Implement the idea in https://github.com/tendermint/tendermint/issues/2386
* Add synchronization in lite verify 
* Change `Certify` to `Verify`

### CODE REVIEW

File: lite/dynamic_verifier.go [hunk 1]
```diff
@@ -56,7 +65,38 @@ func (ic *DynamicVerifier) ChainID() string {
...
+			ic.mtx.Lock()
+			delete(ic.pendingVerifications, shdr.Height)
+			ic.mtx.Unlock()
+		}()
+		ic.mtx.Unlock()
+	}
+	//Get the exact trusted commit for h, and if it is
+	// equal to shdr, then don't even verify it,
+	// and just return nil.
+	trustedFCSameHeight, err := ic.trusted.LatestFullCommit(ic.chainID, shdr.Height, shdr.Height)
```
Comment (cwgoes) [inline lite/dynamic_verifier.go]: Is this stored in RAM or on disk?

Comment (HaoyangLiu) [reply lite/dynamic_verifier.go]: Both of them will store the commits. Here we create disk db and memory db to cache the verified commits:
https://github.com/tendermint/tendermint/blob/d419fffe18531317c28c29a292ad7d253f6cafdf/lite/proxy/verifier.go#L16

Actually, I have changed the [test](https://github.com/HaoyangLiu/tendermint/blob/5f1af3961e7a65bae5515c338fbb9630e24b80ba/lite/dynamic_verifier_test.go#L130) to cover the new code. Next I will create a new test to simulate concurrent requests.

**Comment by jaekwon** [comment]
Very cool.  Has this been tested HaoyangLiu?  What are the next steps?

**Comment by HaoyangLiu** [comment]
@jaekwon 
Sorry for delayed replay. Recently I was so busy with IRISNET develop. I have add a small [test](https://github.com/HaoyangLiu/tendermint/blob/5f1af3961e7a65bae5515c338fbb9630e24b80ba/lite/dynamic_verifier_test.go#L130) the cover the new code. Next I will add more test and implement an LRU cache to improve performance.

**Comment by HaoyangLiu** [comment]
@cwgoes 
Currently, two trust store will be created for tendermint lite: [memory cache store](https://github.com/tendermint/tendermint/blob/cd172acee82aa2aa9c34735eb8f65827c943b623/lite/proxy/verifier.go#L16) and db store. The memory cache maximum size is 10. Mayge we should parameterize the size. The trust memory cache will only keep 10 of the highest trust commit, please refer to the [cache deleting algorithm](
https://github.com/tendermint/tendermint/blob/cd172acee82aa2aa9c34735eb8f65827c943b623/lite/dbprovider.go#L188). Strictly speaking, though it is not the LRU cache algorithm, I think it is also reasonable and in many cases it is equivalent to LRU algorithm. For instances, in most user cases, user will query the latest blockchain state, such as account information. Then latest trust commits will be cached by memory cache store. The LRU cache algorithm will do the same operation. 
In summary, I think the current memory cache store can meet our requirements. But there is  an improvement point: parameterizing the memory cache size. 
How do you think about this? @cwgoes

**Comment by HaoyangLiu** [comment]
@cwgoes @jaekwon 
I have add a test named `TestConcurrencyInquirerVerify` to verify this changes.

**Comment by cwgoes** [comment]
> Currently, two trust store will be created for tendermint lite: memory cache store and db store. The memory cache maximum size is 10. Mayge we should parameterize the size.

This sounds OK, parameterize cache size is definitely a good idea.

File: lite/proxy/verifier.go [hunk 2]
```diff
@@ -8,12 +8,12 @@ import (
 	lclient "github.com/tendermint/tendermint/lite/client"
 )
 
-func NewVerifier(chainID, rootDir string, client lclient.SignStatusClient, logger log.Logger) (*lite.DynamicVerifier, error) {
+func NewVerifier(chainID, rootDir string, client lclient.SignStatusClient, logger log.Logger, cacheSize int) (*lite.DynamicVerifier, error) {
```
Comment (xla) [inline lite/proxy/verifier.go]: This signature is getting quite large, would benefit from [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis), an example from another part of the codebase:

https://github.com/tendermint/tendermint/blob/0c9c3292c918617624f6f3fbcd95eceade18bcd5/privval/socket.go#L41-L54

**Comment by HaoyangLiu** [comment]
@xla 
Thanks for your feedback. I will modify the code as your suggestions.

File: lite/proxy/verifier.go (hunk 2 above)
Comment (HaoyangLiu) [reply lite/proxy/verifier.go]: @xla 
I think function options may be not suitable for `NewVerifier`. Because `NewVerifier` use these parameters for processing network operation, not just variable assignment.
